app/src/main/java/org/kiwix/kiwixmobile/nav/destination/library/OnlineLibraryFragment.kt

```diff
@@ -301,30 +309,74 @@ class OnlineLibraryFragment : BaseFragment(), FragmentActivityExtensions {
     }
   }
 
-  private fun onBookItemClick(item: LibraryListItem.BookItem) {
-    when {
-      isNotConnected -> {
-        noInternetSnackbar()
-        return
+  private fun checkExternalStorageWritePermission(): Boolean {
+    return hasPermission(WRITE_EXTERNAL_STORAGE).also { permissionGranted ->
+      if (!permissionGranted) {
+        if (shouldShowRationale(WRITE_EXTERNAL_STORAGE)) {
+          alertDialogShower.show(
+            KiwixDialog.WriteStoragePermissionRationale,
+            ::requestExternalStoragePermission
+          )
+        } else {
+          requestExternalStoragePermission()
+        }
       }
-      noWifiWithWifiOnlyPreferenceSet -> {
-        dialogShower.show(WifiOnly, {
-          sharedPreferenceUtil.putPrefWifiOnly(false)
-          downloadFile(item.book)
-        })
-        return
+    }
+  }
+
+  private fun requestExternalStoragePermission() {
+    ActivityCompat.requestPermissions(
+      requireActivity(), arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE),
+      REQUEST_STORAGE_PERMISSION
+    )
+  }
+
+  private fun shouldShowRationale(writeExternalStorage: String) =
+    ActivityCompat.shouldShowRequestPermissionRationale(requireActivity(), writeExternalStorage)
+
+  override fun onRequestPermissionsResult(
+    requestCode: Int,
+    permissions: Array<out String>,
+    grantResults: IntArray
+  ) {
+    super.onRequestPermissionsResult(requestCode, permissions, grantResults)
+    if (requestCode == REQUEST_STORAGE_PERMISSION &&
+      grantResults.isNotEmpty() &&
+      grantResults[0] != PERMISSION_GRANTED
```

Comment:
We should check here the result of the first permission. `PERMISSION_GRANTED`doesn't give us explicitly which permission we asked for. It should be like this.
```kt
if (requestCode == REQUEST_STORAGE_PERMISSION && permissions.isNotEmpty() && permissions[0] != Manifest.permission.WRITE_EXTERNAL_STORAGE)
```

Reply:
done